samples/fpgaperf_counter/fpgaperf_counter.c

```diff
@@ -0,0 +1,639 @@
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#define _GNU_SOURCE
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <glob.h>
+#include <libudev.h>
+#include <stdint.h>
+#include <errno.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <sys/ioctl.h>
+#include <linux/perf_event.h>
+#include <linux/hw_breakpoint.h>
+#include <asm/unistd.h>
+#include <opae/log.h>
+#include "opae_int.h"
+#include <opae/properties.h>
+#include <opae/utils.h>
+#include <opae/fpga.h>
+#include <regex.h>
+#include "fpgaperf_counter.h"
+
+
+#define DFL_PERF_FME 		"/sys/bus/pci/devices/*%x*:*%x*:*%x*.*%x*/fpga_region/region*/dfl-fme.*"
+
+#define DFL_PERF_SYSFS		"/sys/bus/event_source/devices/dfl_fme"
+
+#define DFL_PERF_STR_MAX	256
+
+#define PERF_EVNET_PATTERN	"event=(0x[0-9a-fA-F]{2}),evtype=(0x[0-9a-fA-F]{2}),portid=(0x[0-9a-fA-F]{2})"
+
+#define PERF_CONFIG_PATTERN	"config:([0-9]{1,})-([0-9]{2,})"
+
+#define PARSE_MATCH_INT(_p, _m, _v, _b)				\
+	do {							\
+		errno = 0;					\
+		_v = strtoul(_p + _m, NULL, _b);		\
+		if (errno) {					\
+			OPAE_MSG("error parsing int");		\
+		}						\
+	} while (0)
+
+/* mutex to protect fpga perf pmu data structures */
+pthread_mutex_t fpga_perf_lock = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
+
+/* Read format structure*/
+struct read_format {
+	uint64_t nr;
+	struct {
+		uint64_t value;
+		uint64_t id;
+	} values[];
+};
+
+typedef struct  {
+	char event_name[DFL_PERF_STR_MAX];
+	uint64_t config;
+	int fd;
+	uint64_t id;
+	uint64_t start_value;
+	uint64_t stop_value;
+} perf_event_type;
+
+typedef struct {
+	char format_name[DFL_PERF_STR_MAX];
+	uint64_t shift;
+} perf_format_type;
+
+typedef struct {
+	char dfl_fme_name[DFL_PERF_STR_MAX];
+	uint64_t type;
+	uint64_t cpumask;
+	uint64_t num_format;
+	perf_format_type *format_type;
+	uint64_t num_perf_events;
+	perf_event_type *perf_events;
+} fpga_perf_counter;
+
+/* Not static so other tools can access the PMU data */
+fpga_perf_counter *g_fpga_perf = NULL;
+
```

Comment:
If other tools need to access PMU data then maybe we should provide API functions to get them.
Also, are we assuming we can only have one of these at a time? What if I have multiple cards and I have an application what wants to monitor perf data in multiple threads?